### ISSUE TITLE
CF: incorrect warnings for variable attributes scale_factor and add_offset

### ISSUE THREAD

**Issue by mhidas** [ioos/compliance-checker#842]
In #646 a new [check](https://github.com/ioos/compliance-checker/pull/646/files#diff-f14722526b669b2144710ff4feeff8d4R419) was added to the CF suite to check that certain variable attributes have the same data type as their parent variable. I'm pretty sure the attributes `scale_factor` and `add_offset` should **not** be included in this check.

The [NetCDF User Guide](https://www.unidata.ucar.edu/software/netcdf/docs/attribute_conventions.html) (referenced in the doc string) says
> When scale_factor and add_offset are used for packing, the associated variable (containing the packed data) is typically of type byte or short, whereas the unpacked values are intended to be of type float or double. The attributes scale_factor and add_offset should both be of the type intended for the unpacked data, e.g. float or double.

This means in general these two attributes (type float/double) will be of a different type to the variable (byte/short).

CF section 8.1 [Packed Data](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#packed-data) restricts this somewhat, but still allows the attributes to have different data type than the variable.
> However, if the scale_factor and add_offset attributes are of a different data type from the variable (containing the packed data) then the unpacked data should match the type of these attributes, which must both be of type float or both be of type double . An additional restriction in this case is that the variable containing the packed data must be of type byte , short or int . It is not advised to unpack an int into a float as there is a potential precision loss.

**Comment by mhidas** [comment]
Sorry I missed this earlier. Haven't had much time to look at checker issues in the past year. We're finally looking to upgrade to version 4.3.x (still using 4.1.1. now), so now we're testing it against various IMOS data.

**Comment by daltonkell** [comment]
Thanks for pointing this out, @mhidas. So really, the check should be specifically for `scale_factor` and `add_offset`:

```
if (scale_factor_dtype == packed_var_dtype) and (add_offset_dtype == packed_var_dtype):
    continue # all good
else:
    # check dtype of scale_factor and add_offset must be both byte or float
```

and repeat for `add_offset`. I think this is more accurate to the specification. Any thoughts?

**Comment by mhidas** [comment]
Yes, something like that, except in the `else` clause above, you should
* check that `scale_factor` and `add_offset` are both of type **float** or both be of type **double**; and also
* check that the variable with these attributes is of type byte, short or int

**Comment by benjwadams** [comment]
Packed coordinates look to be checked by `check_packed_data`: https://github.com/ioos/compliance-checker/blob/master/compliance_checker/cf/cf.py#L4130-L4236

I think we'd probably be covered by just removing the attributes from the list to be checked, and instead use this method which already works.

**Comment by mhidas** [comment]
:+1: Yep, that makes sense.

One minor comment on the existing code 
https://github.com/ioos/compliance-checker/blob/250f0b4a14ab8ffc034c342830b287758b660275/compliance_checker/cf/cf.py#L4183-L4190

I'm not sure that `np.int` and `np.int64` should be included there. I think both of these are 64-bit types, which maps to the [netCDF type](https://www.unidata.ucar.edu/software/netcdf/docs/data_type.html) `NC_INT64`, i.e. not "byte, short or int"?

**Comment by benjwadams** [comment]
Revisiting some issues here.  Per the previous comment, the CF Conventions state (http://cfconventions.org/cf-conventions/cf-conventions.html#_data_types): 

> Data variables must be one of the following data types: string, char, byte, unsigned byte, short, unsigned short, int, unsigned int, int64, unsigned int64, float or real, and double

This looks like it's CF-1.9 draft.  In any case, the tail issue looks like it'd make more sense as a separate issue.

**Comment by mhidas** [comment]
I've just been testing the version of the code currently on master branch, and for files that contain these variable attributes, I get an exception:
```
The checker throws an exception when checking files that have these attributes:

--------------------------------------------------------------------------------
                         IOOS Compliance Checker Report                         
                         Version 4.3.4.dev103+ge14be12                          
                     Report generated 2021-07-29T06:19:56Z                      
                                     cf:1.6                                     
       http://cfconventions.org/cf-conventions/v1.6.0/cf-conventions.html       
--------------------------------------------------------------------------------
All tests passed!
WARNING: The following exceptions occurred during the cf:1.6 checker (possibly indicate compliance checker issues):
cf:1.6.check_add_offset_scale_factor_type: 'CF1_6Check' object has no attribute '_check_scale_factor_add_offset'
```

**Comment by benjwadams** [comment]
@mhidas, OK to close with functionality in new release?

**Comment by mhidas** [comment]
> @mhidas, OK to close with functionality in new release?

@benjwadams I've just checked with the 4.3.4 release. I no longer get any incorrect warnings about `scale_factor` and `add_offset` attributes, so the original issue reported here seems to be resolved.

However, I do get the exception reported [above](https://github.com/ioos/compliance-checker/issues/842#issuecomment-888861214) for every file that has these attributes. Do you want me to open a separate issue for that, or just rename this one?

Note I am using the CF-1.6 check suite, as that is the version our files should be compliant with.

**Comment by benjwadams** [comment]
@mhidas, as I don't have your source files, has the merged PR above fixed the exceptions?

**Comment by mhidas** [comment]
:+1: Yes it has. Thanks @benjwadams

**Comment by mhidas** [comment]
... However, testing the new 5.0.0 release (the first one to include the above merged PR), I think there's still something not quite right in the logic of this check. I have a variable like this:
```
        short sea_surface_temperature(time, lat, lon) ;
                sea_surface_temperature:_FillValue = -32768s ;
                sea_surface_temperature:add_offset = 291.458831787109 ;
                sea_surface_temperature:scale_factor = 0.00999999977648258 ;
                sea_surface_temperature:valid_min = -32767s ;
                sea_surface_temperature:valid_max = 32767s ;
```
I.e. the variable is of type **short**, and `scale_factor` and `add_offset` are of type **double**
In the checker output, I get
```
§8.1 Packed Data
* Variable sea_surface_temperature and add_offset must be equivalent data types or sea_surface_temperature must be of type byte, short, or int and add_offset must be float or double
* Variable sea_surface_temperature and scale_factor must be equivalent data types or sea_surface_temperature must be of type byte, short, or int and scale_factor must be float or double
```

The complete file is available [here](https://s3-ap-southeast-2.amazonaws.com/imos-data/IMOS/SRS/SST/ghrsst/L3S-1d/dn/2019/20190701092000-ABOM-L3S_GHRSST-SSTfnd-AVHRR_D-1d_dn.nc)

**Comment by benjwadams** [comment]
Thanks for the link to the file @mhidas.  In #906 I fix a problem with comparing short/int types against floating scale_factor/add_offset attributes.  After the fixes, the file no longer raises these compliance checker errors.

**Comment by benjwadams** [comment]
Closing after merged PR.  Feel free to reopen if there are other outstanding issues in the implementation.
